### PR TITLE
[DBGHELP_APITEST] Add test for SymRegisterCallbackW64.

### PR DESCRIPTION
**Pull request by andy-123**
Adds a test for SymRegisterallBackW64.

Shows that the callback given to SymRegisterallBackW64 is not called by ROS.

### CODE REVIEW

**Comment by andy-123** [comment]
Thank you for reviewing / testing. 
I replaced the original dbghelp.dll. 
I did a quick test that shows no crash on w2k3 (with original dbghelp.dll). Maybe is miss some w2k3 updates. I will have a closer look at it.

**Comment by learn-more** [comment]
What might help diagnosing this is printing the full path to `dbghelp.dll` (Using `GetModuleFileNameA`),
and printing the dll version.
See the function `get_module_version` for an example: 
https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/apitests/apphelp/data.c;hb=b2dcb8e0d8925fc1aed2adeaf74ad0291e29f8f6#l803

**Comment by learn-more** [comment]
`0xc06d007f` looks like a delay-load exception, so probably the function is not found.
You can handle this gracefully from a `__pfnDliFailureHook2` callback, so that you'll only skip this test if the function is not found.
Let me know if you want me to commit the required code to the PR.

**Comment by andy-123** [comment]
You're right. The function does not exists in this dll.
Please commit the required code to the PR. Thank you.

**Comment by learn-more** [comment]
I'll squash your commits into one, then leave my commit separate, so it can be used for reference in the case of delay-load failures.

**Comment by andy-123** [comment]
Ok, thank you.

File: modules/rostests/apitests/dbghelp/pdb.c [hunk 1]
```diff
@@ -57,6 +58,118 @@ static void deinit_sym()
     ((PSYMBOL_INFO)(buff))->MaxNameLen = MAX_SYM_NAME; \
 } while (0)
 
+/* modified copy of function from apitests/apphelp/apitest.c */
+BOOL get_module_version(
+    IN HMODULE mod,
```
Comment (tkreuzer) [inline modules/rostests/apitests/dbghelp/pdb.c]: SAL?

File: modules/rostests/apitests/dbghelp/pdb.c [hunk 2]
```diff
@@ -57,6 +58,118 @@ static void deinit_sym()
...
+    OUT VS_FIXEDFILEINFO *fileinfo)
+{
+    BOOL res = FALSE;
+    HRSRC hResInfo = FindResource(mod, MAKEINTRESOURCE(VS_VERSION_INFO), RT_VERSION);
+    DWORD dwSize = SizeofResource(mod, hResInfo);
```
Comment (tkreuzer) [inline modules/rostests/apitests/dbghelp/pdb.c]: No error handling?

File: modules/rostests/apitests/dbghelp/pdb.c [hunk 3]
```diff
@@ -57,6 +58,118 @@ static void deinit_sym()
...
+
+    if (fileinfo == NULL)
+        return FALSE;
+
+    if (hResInfo && dwSize)
+    {
+        UINT uLen;
+
+        HGLOBAL hResData = LoadResource(mod, hResInfo);
+        LPVOID pRes = LockResource(hResData);
```
Comment (tkreuzer) [inline modules/rostests/apitests/dbghelp/pdb.c]: Error handling?

File: modules/rostests/apitests/dbghelp/pdb.c [hunk 4]
```diff
@@ -57,6 +58,118 @@ static void deinit_sym()
...
+        HLOCAL pResCopy = LocalAlloc(LMEM_FIXED, dwSize);
```
Comment (tkreuzer) [inline modules/rostests/apitests/dbghelp/pdb.c]: s.a.

File: modules/rostests/apitests/dbghelp/pdb.c [hunk 5]
```diff
@@ -57,6 +58,118 @@ static void deinit_sym()
...
+            /* Redirect execution to the stub */
+            return (FARPROC)SymRegisterCallbackW64_Stub;
+        }
+    }
+    /* This is not the function you are looking for, continue default behavior (throw exception) */
+    return NULL;
+}
+
+/* Register the failure hook using the magic name '__pfnDliFailureHook2'. */
+PfnDliHook __pfnDliFailureHook2 = DliFailHook;
```
Comment (tkreuzer) [inline modules/rostests/apitests/dbghelp/pdb.c]: What is this good for?

File: modules/rostests/apitests/dbghelp/pdb.c (hunk 1 above)
Comment (andy-123) [reply modules/rostests/apitests/dbghelp/pdb.c]: What is SAL standing for?

Comment (binarymaster) [reply modules/rostests/apitests/dbghelp/pdb.c]: https://reactos.org/wiki/Abbreviations#S

> SAL -- Source Annotation Language, Standard Annotation Language

File: modules/rostests/apitests/dbghelp/pdb.c (hunk 5 above)
Comment (andy-123) [reply modules/rostests/apitests/dbghelp/pdb.c]: Not all dbghelp.dll version have SymRegisterCallbackW64.
This hook gets called if the delayed loaded import for SymRegisterCallBackW64 fails. In this case we can return a stub and prevent an app crash.

File: modules/rostests/apitests/dbghelp/pdb.c (hunk 1 above)
Comment (andy-123) [reply modules/rostests/apitests/dbghelp/pdb.c]: SAL ... I assume because of the IN / OUT.  I find it helpful - why not? Should i remove it?

Comment (ThFabba) [reply modules/rostests/apitests/dbghelp/pdb.c]: It's supposed to be `_In_`

Comment (andy-123) [reply modules/rostests/apitests/dbghelp/pdb.c]: Ah, ok i will fix it. Thanks.

File: modules/rostests/apitests/dbghelp/pdb.c (hunk 5 above)
Comment (zefklop) [reply modules/rostests/apitests/dbghelp/pdb.c]: Using GetProcAddress and bailing out would be much simpler... Why do we delay-load the DLL we are testing anyway ?

Comment (zefklop) [reply modules/rostests/apitests/dbghelp/pdb.c]: This looks like a ton of code...

**Pull request merged**